metals/src/main/scala/scala/meta/internal/metals/debug/DebugDiscovery.scala

```diff
@@ -162,9 +172,226 @@ class DebugDiscovery(
+    val classes = buildTargetClasses.classesOf(buildTarget)
+    val mains = findMainClasses(textDocument, classes)
+    val allTestCases = discoverTestCases(path)
+
+    val allDistances =
+      calculateMainDistances(textDocument, mains, classes, position) ++
+        calculateTestCaseDistances(allTestCases, position) ++
+        calculateTestSuiteDistances(textDocument, allTestCases, position)
+
+    findClosestTarget(allDistances) match {
+      case Some((_, _, "main", _, mainClass: b.ScalaMainClass)) =>
+        findMainToRun(
+          Map(buildTarget -> List(mainClass)),
+          DebugDiscoveryParams(path.toURI.toString, "run"),
+        )
+      case Some((_, _, "testSuite", _, suiteFqn: String)) =>
+        createTestSuiteParams(buildTarget, suiteFqn)
+      case Some((_, _, "testCase", testName: String, suiteFqn: String)) =>
+        createTestCaseParams(buildTarget, suiteFqn, testName)
+      case _ =>
+        Future.failed(NoRunOptionException)
+    }
+  }
+
+  private def findMainClasses(
+      textDocument: TextDocument,
+      classes: BuildTargetClasses.Classes,
+  ): Seq[b.ScalaMainClass] = {
+    val mains = for {
+      occurrence <- textDocument.occurrences
+      if occurrence.role.isDefinition || occurrence.symbol == "scala/main#"
+      symbol = occurrence.symbol
+      mainClass <- {
+        val normal = classes.mainClasses.get(symbol)
+        val fromAnnot = DebugDiscovery
+          .mainFromAnnotation(occurrence, textDocument)
+          .flatMap(classes.mainClasses.get(_))
+        List(normal, fromAnnot).flatten
+      }
+    } yield mainClass
+
+    if (mains.nonEmpty) mains
+    else DebugDiscovery.syntheticMains(textDocument, classes)
+  }
+
+  private def discoverTestCases(
+      path: AbsolutePath
+  ): List[(String, List[TestCaseEntry])] = {
+    import scala.jdk.CollectionConverters._
+    import scala.meta.internal.metals.testProvider.TestExplorerEvent.AddTestCases
+
+    val testUpdates = testProvider.discoverTests(Some(path))
+    testUpdates
+      .flatMap(_.events.asScala)
+      .collect { case AddTestCases(fqn, _, testCases) =>
+        (fqn, testCases.asScala.toList)
+      }
+      .toList
+  }
+
+  private def calculateMainDistances(
+      textDocument: TextDocument,
+      mains: Seq[b.ScalaMainClass],
+      classes: BuildTargetClasses.Classes,
+      position: Position,
+  ): Seq[(Int, Int, String, String, Any)] = {
+    mains.flatMap { mainClass =>
+      val occOpt = textDocument.occurrences.find { occ =>
+        occ.role.isDefinition && (
+          classes.mainClasses.get(occ.symbol).contains(mainClass) ||
+            DebugDiscovery
+              .mainFromAnnotation(occ, textDocument)
+              .flatMap(classes.mainClasses.get(_))
+              .contains(mainClass)
+        )
+      }
+
+      val syntheticOccOpt = if (occOpt.isEmpty) {
+        textDocument.symbols
+          .find(_.symbol.contains(mainClass.getClassName.split('.').last))
```

Comment:
Wouldn't this cause a possibly random symbol to be used if the name corresponds to it?

Probably it's better to contruct the symbol with:
`mainClass.getClassName.replace(".", "\\")` <- this should contruct  the symbol (if we assume it's a topelevel one and I think it's ok to assume it here). We then have to find a symbol that ends with # or . and starts with the contructed name.

A bit more how those symbols look like is here https://scalameta.org/docs/semanticdb/guide.html